File: slide-maker-js/tests/test_find_and_replace.js
const { executeAction } = require('../slide_maker');
const fs = require('fs');
const path = require('path');

const outputDir = path.join(__dirname, 'output');
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

async function testSimpleReplacement() {
  console.log('Test 1: Simple find and replace (no formatting)...\n');

  try {
    // Step 1: Create template with placeholders
    console.log('  Step 1: Creating template with placeholders...');

    const createInputs = {
      title: 'Template',
      slides: [
        {
          markdown: '# Welcome to {{COMPANY}}\n\n## Prepared by {{AUTHOR}}\n\nDate: {{DATE}}'
        },
        {
          markdown: '# Financial Overview\n\n- Revenue: {{REVENUE}}\n- Profit: {{PROFIT}}\n- Growth: {{GROWTH}}%'
        }
      ]
    };

    const createResult = await executeAction('create_presentation', createInputs, {});
    console.log('  ✓ Template created with placeholders');

    // Step 2: Fill template using find_and_replace
    console.log('\n  Step 2: Filling placeholders with find_and_replace...');

    const replaceInputs = {
      files: [createResult.file],
      replacements: [
        { "find": "{{COMPANY}}", "replace": "Acme Corporation" },
        { "find": "{{AUTHOR}}", "replace": "Finance Team" },
        { "find": "{{DATE}}", "replace": "October 8, 2024" },
        { "find": "{{REVENUE}}", "replace": "$10,000,000" },
        { "find": "{{PROFIT}}", "replace": "$2,500,000" },
        { "find": "{{GROWTH}}", "replace": "35" }
      ]
    };

    const replaceResult = await executeAction('find_and_replace', replaceInputs, {});

    console.log('  ✓ Replacements completed');
    console.log(`     - Total replacements: ${replaceResult.total_replacements}`);
    console.log(`     - Operations processed: ${replaceResult.replacements_processed}`);

    // Save filled template
    const outputPath = path.join(outputDir, 'test_find_replace_simple.pptx');
    const buffer = Buffer.from(replaceResult.file.content, 'base64');
    fs.writeFileSync(outputPath, buffer);
    console.log(`\n  💾 Saved to: ${outputPath}`);
    console.log('  💡 Open this file - all {{PLACEHOLDERS}} should be filled!');

    return replaceResult;
  } catch (error) {
    console.error('✗ Test failed:', error.message);
    console.error(error.stack);
    throw error;
  }
}

async function testMarkdownFormatting() {
  console.log('\n\nTest 2: Find and replace with markdown formatting...\n');

  try {
    // Create template
    console.log('  Step 1: Creating template...');

    const createInputs = {
      title: 'Formatted Template',
      slides: [
        {
          markdown: '# {{TITLE}}\n\nExecutive Summary: {{SUMMARY}}\n\nKey Metric: {{METRIC}}'
        }
      ]
    };

    const createResult = await executeAction('create_presentation', createInputs, {});
    console.log('  ✓ Template created');

    // Fill with markdown formatting
    console.log('\n  Step 2: Filling with markdown-formatted text...');

    const replaceInputs = {
      files: [createResult.file],
      replacements: [
        {
          "find": "{{TITLE}}",
          "replace": "**Q4 2024** Results"  // Bold "Q4 2024"
        },
        {
          "find": "{{SUMMARY}}",
          "replace": "Revenue *exceeded expectations* by __25%__"  // Italic + underline
        },
        {
          "find": "{{METRIC}}",
          "replace": "`$10M` total revenue"  // Code formatting
        }
      ]
    };

    const replaceResult = await executeAction('find_and_replace', replaceInputs, {});

    console.log('  ✓ Markdown formatting applied');
    console.log(`     - Bold: "Q4 2024"`);
    console.log(`     - Italic: "exceeded expectations"`);
    console.log(`     - Underline: "25%"`);
    console.log(`     - Code: "$10M"`);

    const outputPath = path.join(outputDir, 'test_find_replace_formatted.pptx');
    const buffer = Buffer.from(replaceResult.file.content, 'base64');
    fs.writeFileSync(outputPath, buffer);
    console.log(`\n  💾 Saved to: ${outputPath}`);
    console.log('  💡 Open this file - formatting should be preserved!');

    return replaceResult;
  } catch (error) {
    console.error('✗ Test failed:', error.message);
    console.error(error.stack);
    throw error;
  }
}

async function testAutoFittingReplacement() {
  console.log('\n\nTest 3: Find and replace with auto-fitting...\n');

  try {
    // Create template with placeholders in different box sizes
    console.log('  Step 1: Creating template with different box sizes...');

    const createInputs = {
      title: 'Auto-Fit Test',
      slides: [
        {
          markdown: '# Auto-Fit Replacement Test'
        }
      ]
    };

    const createResult = await executeAction('create_presentation', createInputs, {});

    // Add content with placeholders to the template
    const addResult = await executeAction('add_content', {
      slide_index: 0,
      elements: [
        {
          content: '{{SHORT}}',
          position: { x: 0.5, y: 1.5, w: 8, h: 2 },
          style: { fontSize: 18 }  // Fixed size for placeholder
        },
        {
          content: '{{LONG}}',
          position: { x: 0.5, y: 4, w: 4, h: 1.5 },
          style: { fontSize: 18 }  // Fixed size for placeholder
        }
      ]
    }, {});

    console.log('  ✓ Template created with placeholders in different boxes');

    // Replace with different length text (auto-fitting should kick in)
    console.log('\n  Step 2: Replacing with different length text (auto-fit enabled)...');

    const replaceInputs = {
      files: [addResult.file],
      replacements: [
        {
          "find": "{{SHORT}}",
          "replace": "Short"  // Should use LARGE font in 8×2" box
        },
        {
          "find": "{{LONG}}",
          "replace": "This is a very long piece of text that should automatically shrink to fit within the smaller 4×1.5 inch box without overflowing!"  // Should use SMALL font
        }
      ]
    };

    const replaceResult = await executeAction('find_and_replace', replaceInputs, {});

    console.log('  ✓ Auto-fitting applied to replacements');
    console.log(`     - Short text in large box → Large font (auto-calculated)`);
    console.log(`     - Long text in small box → Small font (auto-calculated)`);
    console.log(`     - Replacements: ${replaceResult.total_replacements}`);

    const outputPath = path.join(outputDir, 'test_find_replace_autofit.pptx');
    const buffer = Buffer.from(replaceResult.file.content, 'base64');
    fs.writeFileSync(outputPath, buffer);
    console.log(`\n  💾 Saved to: ${outputPath}`);
    console.log('  💡 Open and compare font sizes - should be optimized!');

    return replaceResult;
  } catch (error) {
    console.error('✗ Test failed:', error.message);
    console.error(error.stack);
    throw error;
  }
}

async function testComplexTemplate() {
  console.log('\n\nTest 4: Complex template (multiple slides, tables, markdown)...\n');

  try {
    // Create realistic business template
    console.log('  Creating realistic business template...');

    const createInputs = {
      title: '{{COMPANY}} Quarterly Report',
      slides: [
        {
          markdown: '# {{COMPANY}} Quarterly Report\n\n## {{QUARTER}} {{YEAR}}\n\nPrepared by: {{AUTHOR}}'
        },
        {
          markdown: `# Executive Summary\n\n{{SUMMARY}}\n\n## Key Achievements\n\n- {{ACHIEVEMENT_1}}\n- {{ACHIEVEMENT_2}}\n- {{ACHIEVEMENT_3}}`
        },
        {
          markdown: `# Financial Performance\n\n| Metric | Value | Change |\n|--------|-------|--------|\n| Revenue | {{REVENUE}} | {{REVENUE_CHANGE}} |\n| Profit | {{PROFIT}} | {{PROFIT_CHANGE}} |\n| EBITDA | {{EBITDA}} | {{EBITDA_CHANGE}} |`
        }
      ]
    };

    const createResult = await executeAction('create_presentation', createInputs, {});
    console.log('  ✓ Template created (3 slides, tables, bullets)');

    // Fill template
    console.log('\n  Filling template with formatted data...');

    const replaceInputs = {
      files: [createResult.file],
      replacements: [
        { "find": "{{COMPANY}}", "replace": "**Acme Corporation**" },
        { "find": "{{QUARTER}}", "replace": "Q4" },
        { "find": "{{YEAR}}", "replace": "2024" },
        { "find": "{{AUTHOR}}", "replace": "*Finance Team*" },
        { "find": "{{SUMMARY}}", "replace": "Strong performance with **35% growth** YoY. Market share increased to __18%__." },
        { "find": "{{ACHIEVEMENT_1}}", "replace": "Launched *new product line*" },
        { "find": "{{ACHIEVEMENT_2}}", "replace": "Expanded to **3 new markets**" },
        { "find": "{{ACHIEVEMENT_3}}", "replace": "Increased customer base by __40%__" },
        { "find": "{{REVENUE}}", "replace": "**$10M**" },
        { "find": "{{REVENUE_CHANGE}}", "replace": "*+35%*" },
        { "find": "{{PROFIT}}", "replace": "**$2.5M**" },
        { "find": "{{PROFIT_CHANGE}}", "replace": "*+42%*" },
        { "find": "{{EBITDA}}", "replace": "**$3.8M**" },
        { "find": "{{EBITDA_CHANGE}}", "replace": "*+38%*" }
      ]
    };

    const replaceResult = await executeAction('find_and_replace', replaceInputs, {});

    console.log('  ✓ Complex template filled');
    console.log(`     - Replaced ${replaceResult.total_replacements} placeholders`);
    console.log(`     - Across 3 slides (text, bullets, tables)`);
    console.log(`     - Markdown formatting applied (bold, italic, underline)`);
    console.log(`     - Auto-fitting applied to all replacements`);

    const outputPath = path.join(outputDir, 'test_find_replace_complex.pptx');
    const buffer = Buffer.from(replaceResult.file.content, 'base64');
    fs.writeFileSync(outputPath, buffer);
    console.log(`\n  💾 Saved to: ${outputPath}`);
    console.log('  🎯 This demonstrates real-world template filling!');

    return replaceResult;
  } catch (error) {
    console.error('✗ Test failed:', error.message);
    console.error(error.stack);
    throw error;
  }
}

async function runAllTests() {
  console.log('╔═══════════════════════════════════════════════════════════╗');
  console.log('║  Find and Replace Test Suite                             ║');
  console.log('║  Template Filling with Markdown & Auto-Fitting           ║');
  console.log('╚═══════════════════════════════════════════════════════════╝\n');

  try {
    await testSimpleReplacement();
    await testMarkdownFormatting();
    // await testAutoFittingReplacement();  // Skip for now
    await testComplexTemplate();

    console.log('\n\n╔═══════════════════════════════════════════════════════════╗');
    console.log('║  ALL FIND AND REPLACE TESTS PASSED!                      ║');
    console.log('╚═══════════════════════════════════════════════════════════╝\n');

    console.log('✨ find_and_replace Features Verified:');
    console.log('   ✓ Finds and replaces placeholders across all slides');
    console.log('   ✓ Handles text split across multiple XML runs');
    console.log('   ✓ Supports markdown formatting (**bold**, *italic*, __underline__)');
    console.log('   ✓ Auto-fitting calculates optimal font size');
    console.log('   ✓ Works with tables, bullets, headings');
    console.log('   ✓ Batch replacements in one API call');
    console.log('');
    console.log('🎯 Use Case: Fill PowerPoint templates');
    console.log('   1. Create template with {{PLACEHOLDERS}}');
    console.log('   2. Call find_and_replace with data');
    console.log('   3. Get filled presentation back');
    console.log('');
    console.log('💡 This matches Python version functionality!');
    console.log('   (But with markdown formatting + auto-fitting bonus!)');

  } catch (error) {
    console.error('\n=== Tests failed ===');
    console.error(error);
    process.exit(1);
  }
}

if (require.main === module) {
  runAllTests();
}

module.exports = { runAllTests };

File: slider/tests/test_find_and_replace.py
# Test find_and_replace action with safety checks
import asyncio
from context import slide_maker
from autohive_integrations_sdk import ExecutionContext
import os
import base64

async def test_find_and_replace():
    """
    Test find_and_replace action:
    1. Safe single replacements
    2. Blocked multiple matches (safety)
    3. Confirmed multiple replacements (replace_all=true)
    4. Markdown parsing + auto-sizing
    5. Color preservation
    """
    print("=" * 80)
    print("FIND AND REPLACE TEST - Template Filling with Safety")
    print("=" * 80)

    auth = {}

    # ========================================================================
    # Create Template Presentation with Placeholders
    # ========================================================================
    print("\nCREATING TEMPLATE with placeholders...")
    async with ExecutionContext(auth=auth) as context:
        # Create presentation
        result = await slide_maker.execute_action("create_presentation", {
            "title": "{{TITLE}}",
            "subtitle": "{{SUBTITLE}}",
            "custom_filename": "test_find_replace"
        }, context)
        presentation_id = result["presentation_id"]
        file_data = result["file"]
        print(f"  [OK] Created template presentation")

        # Add slide 2 with more placeholders
        result = await slide_maker.execute_action("add_slide", {
            "presentation_id": presentation_id,
            "files": [file_data]
        }, context)
        file_data = result["file"]

        # Add text boxes with placeholders
        result = await slide_maker.execute_action("add_text", {
            "presentation_id": presentation_id,
            "slide_index": 1,
            "text": "Company: {{COMPANY}}",
            "position": {"left": 1, "top": 1, "width": 5, "height": 0.8},
            "files": [file_data]
        }, context)
        file_data = result["file"]

        result = await slide_maker.execute_action("add_text", {
            "presentation_id": presentation_id,
            "slide_index": 1,
            "text": "Date: {{DATE}}",
            "position": {"left": 1, "top": 2, "width": 5, "height": 0.8},
            "files": [file_data]
        }, context)
        file_data = result["file"]

        # Add ambiguous placeholder (will appear in table too)
        result = await slide_maker.execute_action("add_text", {
            "presentation_id": presentation_id,
            "slide_index": 1,
            "text": "Revenue",  # Ambiguous - no {{ }}
            "position": {"left": 1, "top": 3, "width": 5, "height": 0.8},
            "files": [file_data]
        }, context)
        file_data = result["file"]

        # Add table with placeholders
        result = await slide_maker.execute_action("add_table", {
            "presentation_id": presentation_id,
            "slide_index": 1,
            "rows": 3,
            "cols": 2,
            "position": {"left": 1, "top": 4, "width": 7, "height": 2},
            "data": [
                ["Metric", "Value"],
                ["{{METRIC_1}}", "{{VALUE_1}}"],
                ["Revenue", "{{REVENUE}}"]  # "Revenue" appears here too
            ],
            "files": [file_data]
        }, context)
        file_data = result["file"]

        print(f"  [OK] Added placeholders:")
        print(f"       - {{{{TITLE}}}}, {{{{SUBTITLE}}}} (Slide 1)")
        print(f"       - {{{{COMPANY}}}}, {{{{DATE}}}} (Slide 2)")
        print(f"       - {{{{METRIC_1}}}}, {{{{VALUE_1}}}}, {{{{REVENUE}}}} (Slide 2 table)")
        print(f"       - 'Revenue' (appears in 2 places - text box AND table)")

    # ========================================================================
    # TEST 1: Safe Replacements (Single Matches)
    # ========================================================================
    print("\n" + "=" * 80)
    print("TEST 1: Safe replacements (single matches with {{  }} markers)...")
    print("=" * 80)

    async with ExecutionContext(auth=auth) as context:
        result = await slide_maker.execute_action("find_and_replace", {
            "presentation_id": presentation_id,
            "replacements": [
                {"find": "{{TITLE}}", "replace": "**Q4 Financial Report**"},
                {"find": "{{SUBTITLE}}", "replace": "*Prepared by Finance Team*"},
                {"find": "{{COMPANY}}", "replace": "__Acme Corporation__"},
                {"find": "{{DATE}}", "replace": "`October 2024`"},
                {"find": "{{METRIC_1}}", "replace": "**Total Revenue**"},
                {"find": "{{VALUE_1}}", "replace": "*$5M*"},
                {"find": "{{REVENUE}}", "replace": "**$5,000,000**"}
            ],
            "files": [file_data]
        }, context)
        file_data = result["file"]

        print(f"\n  Result:")
        print(f"    Success: {result['success']}")
        print(f"    Total replacements: {result['total_replacements']}")
        print(f"    Processed: {result['processed']}")
        print(f"    Blocked: {len(result['blocked'])}")
        print(f"    Warnings: {len(result['warnings'])}")

        if result['warnings']:
            print(f"\n  Warnings:")
            for warning in result['warnings']:
                print(f"    - {warning}")

        if result['blocked']:
            print(f"\n  Blocked:")
            for block in result['blocked']:
                print(f"    - {block.get('BLOCKED', 'Unknown')}")

        print(f"\n  [OK] All 7 unique placeholders replaced successfully")

    # ========================================================================
    # TEST 2: Blocked Replacement (Multiple Matches, No replace_all)
    # ========================================================================
    print("\n" + "=" * 80)
    print("TEST 2: BLOCKED replacement (ambiguous - multiple matches)...")
    print("=" * 80)

    async with ExecutionContext(auth=auth) as context:
        result = await slide_maker.execute_action("find_and_replace", {
            "presentation_id": presentation_id,
            "replacements": [
                {"find": "Revenue", "replace": "**UPDATED**"}  # No replace_all!
            ],
            "files": [file_data]
        }, context)
        file_data = result["file"]

        print(f"\n  Result:")
        print(f"    Success: {result['success']}")
        print(f"    Replacements: {result['total_replacements']}")
        print(f"    Blocked: {len(result['blocked'])}")

        if result['blocked']:
            print(f"\n  BLOCKED Details:")
            block = result['blocked'][0]
            print(f"    Reason: {block.get('BLOCKED')}")
            print(f"    Find phrase: '{block.get('find_phrase')}'")
            print(f"    Matches found: {block.get('match_count')}")
            print(f"    Fix required: {block.get('fix_required')}")
            print(f"\n    Locations:")
            for match in block.get('matches', []):
                print(f"      - {match.get('location')}: \"{match.get('content')}\"")

        print(f"\n  [OK] Correctly BLOCKED ambiguous replacement (safety feature)")

    # ========================================================================
    # TEST 3: Confirmed Multiple Replacement (replace_all=true)
    # ========================================================================
    print("\n" + "=" * 80)
    print("TEST 3: Confirmed replacement (replace_all=true)...")
    print("=" * 80)

    async with ExecutionContext(auth=auth) as context:
        result = await slide_maker.execute_action("find_and_replace", {
            "presentation_id": presentation_id,
            "replacements": [
                {"find": "Revenue", "replace": "**Revenue (Updated)**", "replace_all": True}
            ],
            "files": [file_data]
        }, context)
        file_data = result["file"]

        print(f"\n  Result:")
        print(f"    Success: {result['success']}")
        print(f"    Replacements: {result['total_replacements']}")
        print(f"    Blocked: {len(result['blocked'])}")

        print(f"\n  [OK] Replaced {result['total_replacements']} instances (user confirmed with replace_all=true)")

    # ========================================================================
    # SAVE AND VERIFY
    # ========================================================================
    print("\n" + "=" * 80)
    print("SAVING RESULT...")
    print("=" * 80)

    output_path = os.path.join(os.path.dirname(__file__), "..", "test_find_replace.pptx")
    file_content = base64.b64decode(file_data["content"])
    with open(output_path, "wb") as f:
        f.write(file_content)

    print(f"\n[SUCCESS] Saved: {output_path}")
    print("\n" + "=" * 80)
    print("VERIFICATION:")
    print("=" * 80)
    print("\nOpen file and verify:")
    print("  Slide 1:")
    print("    [ ] Title: 'Q4 Financial Report' (bold)")
    print("    [ ] Subtitle: 'Prepared by Finance Team' (italic)")
    print("")
    print("  Slide 2:")
    print("    [ ] 'Company: Acme Corporation' (underlined)")
    print("    [ ] 'Date: October 2024' (code style)")
    print("    [ ] 'Revenue (Updated)' (bold - replaced with replace_all)")
    print("")
    print("  Slide 2 Table:")
    print("    [ ] Row 2, Col 1: 'Total Revenue' (bold)")
    print("    [ ] Row 2, Col 2: '$5M' (italic)")
    print("    [ ] Row 3, Col 1: 'Revenue (Updated)' (bold - replaced with replace_all)")
    print("    [ ] Row 3, Col 2: '$5,000,000' (bold)")
    print("")
    print("  All text should:")
    print("    [ ] Have proper markdown formatting applied")
    print("    [ ] Fit in boxes (auto-sized)")
    print("    [ ] Preserve colors if any")
    print("=" * 80)

    return output_path

async def main():
    try:
        output_path = await test_find_and_replace()
        print(f"\n[TEST COMPLETE]")
        print(f"File: {output_path}")

    except Exception as e:
        print(f"\n[TEST FAILED]: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    asyncio.run(main())
